### PR TITLE
Vector conversion support

### LINKED ISSUE #9317: Add support for Converts, Widen and Narrow
In support of https://github.com/dotnet/corefx/issues/1605, RyuJIT should support the new instrinsics as proposed.

### PR DESCRIPTION
**Pull request by helloguo**
#9318
 
This is preliminary. Do not merge.

This build is on top of PR#9318. It fixes failures and implements the remaining intrinsics.

@sivarv @russellhadley PTAL

Fixes #9317

### CODE REVIEW

File: src/jit/simd.cpp [hunk 1]
```diff
@@ -2622,6 +2628,46 @@ GenTreePtr Compiler::impSIMDIntrinsic(OPCODE                opcode,
...
         break;
 
+        case SIMDIntrinsicNarrow:
+        {
+            assert(!instMethod);
+            op2 = impSIMDPopStack(simdType);
+            op1 = impSIMDPopStack(simdType);
+
+            simdTree = gtNewSIMDNode(simdType, op1, op2, simdIntrinsicID, baseType, size);
+            retVal   = simdTree;
```
Comment (sivarv) [inline src/jit/simd.cpp]: It would be a good idea to add a comment here explaining what op1 and op2 is so that one doesn't have to look at managed API to understand this.

File: src/jit/simdintrinsiclist.h [hunk 2]
```diff
@@ -119,6 +119,16 @@ SIMD_INTRINSIC("ConditionalSelect",         false,       Select,
...
+SIMD_INTRINSIC("ConvertToInt64",            false,       ConvertToInt64,           "ConvertToInt64",         TYP_STRUCT,     1,      {TYP_STRUCT, TYP_UNDEF,  TYP_UNDEF},   {TYP_DOUBLE, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
+SIMD_INTRINSIC("ConvertToUInt64",           false,       ConvertToUInt64,          "ConvertToUInt64",        TYP_STRUCT,     1,      {TYP_STRUCT, TYP_UNDEF,  TYP_UNDEF},   {TYP_DOUBLE, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
+SIMD_INTRINSIC("Narrow",                    false,       Narrow,                   "Narrow",                 TYP_STRUCT,     2,      {TYP_STRUCT, TYP_STRUCT, TYP_UNDEF},   {TYP_INT, TYP_DOUBLE, TYP_LONG, TYP_CHAR, TYP_SHORT, TYP_UINT, TYP_ULONG, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
+SIMD_INTRINSIC("Widen",                     false,       Widen,                    "Widen",                  TYP_VOID,       3,      {TYP_STRUCT, TYP_BYREF,  TYP_BYREF},   {TYP_INT, TYP_FLOAT, TYP_CHAR, TYP_UBYTE, TYP_BYTE, TYP_SHORT, TYP_UINT, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
```
Comment (sivarv) [inline src/jit/simdintrinsiclist.h]: Please add comments here explaining Narrow and Widen intrinsics.  Others are self-explanatory.

File: src/jit/simdintrinsiclist.h [hunk 3]
```diff
@@ -134,6 +144,8 @@ SIMD_INTRINSIC("ShiftRightInternal",        false,       ShiftRightInternal,
 SIMD_INTRINSIC("UpperSave",                 false,       UpperSave,                "UpperSave Internal",     TYP_STRUCT,     2,      {TYP_UNDEF, TYP_UNDEF, TYP_UNDEF},     {TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
 SIMD_INTRINSIC("UpperRestore",              false,       UpperRestore,             "UpperRestore Internal",  TYP_STRUCT,     2,      {TYP_UNDEF, TYP_UNDEF, TYP_UNDEF},     {TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
 
+SIMD_INTRINSIC("WidenHi", false, WidenHi, "WidenHi", TYP_VOID, 3, { TYP_UNDEF, TYP_UNDEF,  TYP_UNDEF }, { TYP_INT, TYP_FLOAT, TYP_CHAR, TYP_UBYTE, TYP_BYTE, TYP_SHORT, TYP_UINT, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF })
```
Comment (sivarv) [inline src/jit/simdintrinsiclist.h]: >SIMD_INTRINSIC("WidenHi", false, W [](start = 0, length = 34)

Nit: fields are not properly aligned.

Also please add a comment that this is an internal intrinsic.

Comment (sivarv) [inline src/jit/simdintrinsiclist.h]: Please also add comments explaining its semantics.

---
In reply to: [104520551](https://github.com/dotnet/coreclr/pull/9920#discussion_r104520551) [](ancestors = 104520551)

File: src/jit/simd.cpp [hunk 4]
```diff
@@ -2622,6 +2628,46 @@ GenTreePtr Compiler::impSIMDIntrinsic(OPCODE                opcode,
...
+
+        case SIMDIntrinsicWiden:
+        {
+            GenTree* dstAddrHi = impSIMDPopStack(TYP_BYREF);
+            GenTree* dstAddrLo = impSIMDPopStack(TYP_BYREF);
+            op1                = impSIMDPopStack(simdType);
+            GenTree* dupOp1    = fgInsertCommaFormTemp(&op1, gtGetStructHandleForSIMD(simdType, baseType));
+
+            // Widen the lower half and assign it to dstAddrLo.
+            simdTree = gtNewSIMDNode(simdType, op1, nullptr, simdIntrinsicID, baseType, size);
```
Comment (sivarv) [inline src/jit/simd.cpp]: >simdTree = gtNewSIMDNode(simdType, op1, nullptr, simdIntrinsicID, baseType, size); [](start = 12, length = 82)

We are re-using SIMDIntrinsicWiden to mean two different things depending on the context: prior to importation to stand the intrinsic and after importation to stand for widening the lower-half.

What we are saving is one adding one additional enum, which is not a big saving in IMO.  I would suggest to add SIMDIntrinsicWidenLow to avoid any confusion.

File: src/jit/instrsxarch.h [hunk 5]
```diff
@@ -183,6 +183,7 @@ INST3( movq,        "movq"        , 0, IUM_WR, 0, 0, PCKDBL(0xD6), BAD_CODE, SSE
 INST3( movsdsse2,   "movsd"       , 0, IUM_WR, 0, 0, SSEDBL(0x11), BAD_CODE, SSEDBL(0x10))
 
 INST3( punpckldq,   "punpckldq"   , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, PCKDBL(0x62))
+INST3( punpckhdq,   "punpckhdq"   , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, PCKDBL(0x6A))
```
Comment (sivarv) [inline src/jit/instrsxarch.h]: This one too should be under #ifndef LEGACY_BACKEND? because it is not required by legacy x86 backend.

File: src/jit/instrsxarch.h [hunk 6]
```diff
@@ -334,8 +362,12 @@ INST3( vpbroadcastw, "pbroadcastw" , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, SS
 INST3( vpbroadcastd, "pbroadcastd" , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, SSE38(0x58))   // Broadcast int32 value from reg/memory to entire ymm register
 INST3( vpbroadcastq, "pbroadcastq" , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, SSE38(0x59))   // Broadcast int64 value from reg/memory to entire ymm register
 INST3( vextractf128, "extractf128" , 0, IUM_WR, 0, 0, SSE3A(0x19),  BAD_CODE, BAD_CODE)      // Extract 128-bit packed floating point values
+INST3( vextracti128, "extracti128" , 0, IUM_WR, 0, 0, SSE3A(0x39),  BAD_CODE, BAD_CODE)      // Extract 128-bit packed integer values
 INST3( vinsertf128,  "insertf128"  , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, SSE3A(0x18))   // Insert 128-bit packed floating point values
+INST3( vinserti128,  "inserti128"  , 0, IUM_WR, 0, 0, BAD_CODE,     BAD_CODE, SSE3A(0x38))   // Insert 128-bit packed floating point values
```
Comment (sivarv) [inline src/jit/instrsxarch.h]: >Insert 128-bit packed floating point values [](start = 96, length = 43)

typo: "insert 128-bit packed integer values" instead of floating point values

File: src/jit/lsraxarch.cpp [hunk 7]
```diff
@@ -2693,7 +2693,68 @@ void Lowering::TreeNodeInfoInitSIMD(GenTree* tree)
             break;
 
         case SIMDIntrinsicCast:
+        case SIMDIntrinsicConvertToSingle:
             info->srcCount = 1;
+            if (simdTree->gtSIMDBaseType == TYP_UINT)
+            {
+                info->isInternalRegDelayFree = true;
```
Comment (sivarv) [inline src/jit/lsraxarch.cpp]: >info->isInternalRegDelayFree = true; [](start = 16, length = 36)

In this case as well need internal float regs different from the target reg?

File: src/jit/lsraxarch.cpp [hunk 8]
```diff
@@ -2693,7 +2693,68 @@ void Lowering::TreeNodeInfoInitSIMD(GenTree* tree)
...
+            info->internalIntCount       = 1;
+            if (comp->getSIMDInstructionSet() == InstructionSet_AVX)
+            {
+                info->internalFloatCount = 2;
+            }
+            else
+            {
+                info->internalFloatCount = 1;
+            }
+            if (simdTree->gtSIMDBaseType == TYP_ULONG)
```
Comment (sivarv) [inline src/jit/lsraxarch.cpp]: >           if (simdTree->gtSIMDBaseType == TYP_ULONG) [](start = 1, length = 53)

Nit: please leave a blank line before this if-stmnt for readability.

File: src/jit/codegencommon.cpp [hunk 9]
```diff
@@ -8574,7 +8574,7 @@ void CodeGen::genFnProlog()
     else
 #endif // _TARGET_XARCH_
     {
-        tempMask = initRegs & ~excludeMask & ~regSet.rsMaskResvd;
+        tempMask = initRegs & RBM_ALLINT & ~excludeMask & ~regSet.rsMaskResvd;
```
Comment (sivarv) [inline src/jit/codegencommon.cpp]: > RBM_ALLINT &  [](start = 29, length = 14)

For my benefit please explain why is this change needed?

File: src/jit/simdintrinsiclist.h [hunk 10]
```diff
@@ -119,6 +119,16 @@ SIMD_INTRINSIC("ConditionalSelect",         false,       Select,
 // Cast
 SIMD_INTRINSIC("op_Explicit",               false,       Cast,                     "Cast",                   TYP_STRUCT,     1,      {TYP_STRUCT, TYP_UNDEF,  TYP_UNDEF},   {TYP_INT, TYP_FLOAT, TYP_DOUBLE, TYP_LONG, TYP_CHAR, TYP_UBYTE, TYP_BYTE, TYP_SHORT, TYP_UINT, TYP_ULONG})
 
+// Convert
+SIMD_INTRINSIC("ConvertToSingle",           false,       ConvertToSingle,          "ConvertToSingle",        TYP_STRUCT,     1,      {TYP_STRUCT, TYP_UNDEF,  TYP_UNDEF},   {TYP_INT, TYP_UINT, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF, TYP_UNDEF})
```
Comment (sivarv) [inline src/jit/simdintrinsiclist.h]: >ConvertToSingle [](start = 16, length = 15)

These intrinsic names are not indicative of source type.

So please add comments here describing these intrinsics as well.

File: src/jit/simdcodegenxarch.cpp [hunk 11]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+            inst_RV_RV(INS_vpbroadcastd, tmpReg, tmpReg, targetType, emitActualTypeSize(targetType));
+        }
+        else
+        {
+            inst_RV_RV(INS_movlhps, tmpReg, tmpReg, targetType, emitActualTypeSize(targetType));
+        }
+        inst_RV_RV(INS_orps, targetReg, tmpReg, targetType, emitActualTypeSize(targetType));
+        inst_RV_RV(INS_subps, targetReg, tmpReg, targetType, emitActualTypeSize(targetType));
+        inst_RV_RV(ins, tmpReg2, tmpReg2, targetType, emitActualTypeSize(targetType));
+        inst_RV_RV(INS_addps, targetReg, tmpReg2, targetType, emitActualTypeSize(targetType));
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: Please add comments here what this logic is attempting to achieve.

File: src/jit/simdcodegenxarch.cpp [hunk 12]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+    }
+    else
     {
         inst_RV_RV(ins, targetReg, op1Reg, targetType, emitActualTypeSize(targetType));
     }
     genProduceReg(simdNode);
 }
 
+//----------------------------------------------------------------------------------
+// genSIMDLo64BitConvert: Generate code to convert 1 low-order 64-bit item (long/ulong <-> double)
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: >Generate code to convert 1 low-order 64-bit item (long/ulong <-> double) [](start = 26, length = 72)

I would suggest to rephrase the above comment as follows:

"Generate code to convert lower-most 64-bit item (long/ulong <--> double)

Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: This routine seems to deal with signed long values only.  So comment needs to be updated.



---
In reply to: [104539709](https://github.com/dotnet/coreclr/pull/9920#discussion_r104539709) [](ancestors = 104539709)

File: src/jit/simdcodegenxarch.cpp [hunk 13]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+        // (Note that for mov_xmm2i, the int register is always in the reg2 position)
+        inst_RV_RV(INS_mov_xmm2i, tmpReg, tmpIntReg, TYP_LONG);
+        inst_RV_RV(ins, targetReg, tmpIntReg, baseType, emitActualTypeSize(baseType));
+    }
+    else
+    {
+        inst_RV_RV(ins, tmpIntReg, tmpReg, baseType, emitActualTypeSize(baseType));
+        inst_RV_RV(INS_mov_i2xmm, targetReg, tmpIntReg, TYP_LONG);
+    }
+}
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: This routine will not work for x86 RyuJIT since it has no double registers.

File: src/jit/simdcodegenxarch.cpp [hunk 14]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+    var_types      simdType  = simdNode->TypeGet();
+    regNumber      op1Reg    = genConsumeReg(op1);
+    regNumber      tmpIntReg = genRegNumFromMask(simdNode->gtRsvdRegs & RBM_ALLINT);
+    regNumber      tmpReg;
+    regNumber      tmpReg2;
+    InstructionSet iset = compiler->getSIMDInstructionSet();
+
+    if (iset == InstructionSet_AVX || (baseType == TYP_ULONG))
+    {
+        // For AVX, we need two xmm tmpRegs.
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: >// For AVX, we need two xmm tmpRegs [](start = 8, length = 35)

AVX or basetype is ULONG

File: src/jit/simdcodegenxarch.cpp [hunk 15]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+        tmpReg  = genRegNumFromMask(tmpReg1Mask);
+        tmpReg2 = genRegNumFromMask(tmpRegsMask);
+    }
+    else
+    {
+        if ((intrinsicID == SIMDIntrinsicConvertToDouble) && (targetReg != op1Reg))
+        {
+            // We don't need a tmpReg; we can just use the targetReg.
+            tmpReg = targetReg;
+        }
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: Looks like we don't need an internal register if targetReg != op1Reg.  Right now we are still reserving an internal register in all cases.

Instead we can request Register Allocator to allocate a targetReg different from its operand by marking its op1 as isDelayFree = true.  This will avoid the need for reserving an internal register.

File: src/jit/simdcodegenxarch.cpp [hunk 16]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+        {
+            inst_RV_RV(INS_vpbroadcastq, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        else
+        {
+            inst_RV_RV(INS_movlhps, tmpReg, tmpReg, simdType, emitActualTypeSize(simdType));
+        }
+        inst_RV_RV(INS_orpd, tmpReg2, tmpReg, simdType, emitActualTypeSize(simdType));
+        inst_RV_RV(INS_subpd, tmpReg2, tmpReg, simdType, emitActualTypeSize(simdType));
+        inst_RV_RV(INS_addpd, targetReg, tmpReg2, simdType, emitActualTypeSize(simdType));
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: Please add comments to describe at a high-level what this logic is achieving required conversion.

File: src/jit/simdcodegenxarch.cpp [hunk 17]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+            getEmitter()->emitIns_R_I(rightShiftIns, emitActualTypeSize(simdType), tmpReg2, 8);
+            genSIMDLo64BitConvert(intrinsicID, simdType, baseType, tmpReg2, tmpIntReg, tmpReg2);
+
+            // Now, shift the resulting 64-bits left.
+            getEmitter()->emitIns_R_I(leftShiftIns, emitActualTypeSize(simdType), tmpReg2, 8);
+
+            // Convert v[2], in the lo bits of tmpReg.
+            // For the convert to double, the convert preserves the upper bits in tmpReg2.
+            // For the integer convert, we have to put it in tmpReg and or it in, since movd clears the upper bits.
+            genSIMDLo64BitConvert(intrinsicID, simdType, baseType, tmpReg, tmpIntReg, tmpReg2);
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: >genSIMDLo64BitConvert(intrinsicID, simdType, baseType, tmpReg, tmpIntReg, tmpReg2); [](start = 12, length = 83)

This routine doesn't work for x86 RyuJIT since it assumes 64-bit integer registers.

Please note that we have enabled SIMD even on x86 RyuJIT.

It is fine to stage the work: enable these intrinsics only for 64-bit and file a tracking issue to enable them for x86 RyuJIT.  Subsequently we can do the incremental work required for x86.

File: src/jit/simdcodegenxarch.cpp [hunk 18]
```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
...
+            // We have 8 long elements, 0-3 in op1Reg, 4-7 in op2Reg.
+            // We will generate the following:
+            //   vextracti128 tmpReg, op1Reg, 1       (extract elements 2 and 3 into tmpReg)
+            //   vextracti128 tmpReg2, op2Reg, 1      (extract elements 6 and 7 into tmpReg2)
+            //   vinserti128  tmpReg, tmpReg2, 1       (insert elements 6 and 7 into the high half of tmpReg)
+            //   mov          tmpReg2, op1Reg
+            //   vinserti128  tmpReg2, op2Reg, 1      (insert elements 4 and 5 into the high half of tmpReg2)
+            //   pshufd       tmpReg, tmpReg, XXZX    ( -  - 7L 6L  -  - 3L 2L) in tmpReg
+            //   pshufd       tgtReg, tmpReg2, XXZX   ( -  - 5L 4L  -  - 1L 0L) in tgtReg
+            //   punpcklqdq   tgtReg, tmpReg
```
Comment (sivarv) [inline src/jit/simdcodegenxarch.cpp]: These kind of high-level descriptions would be useful for understanding the logic.

**Comment by helloguo** [comment]
@sivarv Thanks for your feedback. Will address them, test on the latest CoreCLR/CoreFX and make another PR.

**Comment by sivarv** [comment]
Looping in @BruceForstall who did SIMD work for x86. 
@Bruce - FYI, Intel team is trying to add conversion and narrow/widen vector intrinsics.  Initially they will be enabling this for x64 and subsequently for x86.  Right now this is a WIP.  Feel free to review it and provide feedback.

File: tests/dir.props [hunk 19]
```diff
@@ -78,7 +78,8 @@
 
    <!-- list of nuget package sources passed to dnu --> 
    <ItemGroup> 
-     <!-- Need to escape double forward slash (%2F) or MSBuild will normalize to one slash on Unix. -->
+     <!-- Need to escape double forward slash (%2F) or MSBuild will normalize to one slash on Unix. --> 
+     <DnuSourceList Include="C:\Users\win10\fromCarol\corefx\bin\packages\Release" />
```
Comment (BruceForstall) [inline tests/dir.props]: I assume you don't plan to check this in

Comment (helloguo) [inline tests/dir.props]: @BruceForstall You are right, I don't plan check this in.

**Comment by helloguo** [comment]
superseded by https://github.com/dotnet/coreclr/pull/10662
